Fix moderated session presence checking

rosstimothy opened the pull request:
Addresses all of the issues that were preventing presence checking during moderated sessions from working as described in
[#18092](https://github.com/gravitational/teleport/issues/18092#issuecomment-1540900859).

Closes #18092

rosstimothy committed:
Fix moderated session presence checking

Addresses all of the issues that were preventing presence checking during
moderated sessions from working as described in
[#18092](https://github.com/gravitational/teleport/issues/18092#issuecomment-1540900859).

Closes #18092

zmb3 commented on api/types/session_tracker_test.go:
```diff
@@ -0,0 +1,56 @@
+	// Presence is only updated for the matching user
+	require.NoError(t, s.UpdatePresence("fish"))
+	participants := s.GetParticipants()
+	require.Equal(t, clock.Now().UTC(), participants[0].LastActive)
```
How does this work? `UpdatePresence` uses `time.Now`, not the fake clock.

zmb3 commented on lib/client/session.go:
```diff
@@ -492,17 +492,41 @@ func (ns *NodeSession) updateTerminalSize(ctx context.Context, s *tracessh.Sessi
+}
+
+func (s *sessionWriter) Write(p []byte) (int, error) {
+	if s.session.Stdout != nil {
```
When does `Stdout` get set? Does that happen in another goroutine?

rosstimothy committed:
make presence test clearer

rosstimothy committed:
fix presence checking on the web ui

rosstimothy committed:
Refactor web socket message handling

A single message processing loop handles retrieving the envelope
and passing it off to individual message handlers. This allows all
messages to be processed outside of `Read` which was dependent on
the terminal being active to process any messages.

The webauthn challenge response was also moved from a raw message
to a webauthn message. By sending it as a raw message it made
presence checking fail because the response has a `t` in it which
caused the session to be killed during moderated sessions.

rosstimothy committed:
enforce mfa ceremony when joining and cluster wide mfa is enabled

jentfoo reviewed (approved):
Looks reasonable, let me know if you want help testing

rosstimothy committed:
fix conflicts with master

rosstimothy committed:
moderated tests

rosstimothy committed:
Add moderated session tests for the UI

rosstimothy committed:
Add moderated session integration test

rosstimothy committed:
fix lints

rosstimothy committed:
clean up presence test

ravicious commented:
I'm almost finished with reviewing this, though I'm no expert on the code being changed here. I'll try to test this on Monday.

ravicious commented on lib/web/apiserver_test.go:
```diff
@@ -8778,19 +8912,183 @@ func TestModeratedSession(t *testing.T) {
+
+	// run the presence check a few times
+	for i := 0; i < 3; i++ {
+		s.clock.Advance(30 * time.Second)
+		require.NoError(t, waitForOutput(moderatorStream, "Teleport > Please tap your MFA key"))
```
I'm not familiar with Go APIs for reading streams, but is `waitForOutput` guaranteed to receive the expected string even if it's written to the stream between `s.clock.Advance` and `waitForOutput`?

ravicious reviewed (commented):
I tested moderated file transfer with cluster-wide per-session MFA turned on as well as regular file transfer and both work fine after the changes. Moderator presence checks work as expected as well.

I'm yet to review all the tests though.

jakule commented on lib/web/terminal.go:
```diff
@@ -1032,6 +946,18 @@ type WSStream struct {
 
+	// resizeC a channel to forward resize events so that
+	// they happen out of band and don't block reads
+	resizeC chan Envelope
+	// fileTransferRequestC is a channel to facilitate requesting a file transfer
+	fileTransferRequestC chan Envelope
+	// fileTransferDecisionC is a channel to facilitate responding to a file transfer
+	// with an approval or denial
+	fileTransferDecisionC chan Envelope
```
Can we not keep these fields here? WSStream is used by Terminal and Command Handlers. The second one doesn't need them. IMO we should try to find an abstraction that holds only the required fields in a struct.

jakule commented on lib/web/terminal.go:
```diff
@@ -1040,23 +966,27 @@ type WSStream struct {
 	// ws the connection to the UI
 	ws WSConn
+
+	// ctx is used to signal when the terminal session is closing.
+	ctx context.Context
+	// cancel is used to terminate the terminal session.
+	cancel context.CancelFunc
```
https://go.dev/blog/context-and-structs

jakule commented on lib/web/terminal.go:
```diff
@@ -1068,6 +998,199 @@ func (t *WSStream) writeError(err error) error {
+			}
+
+			if ty != websocket.BinaryMessage {
+				t.log.Warnf("Expected binary message, got %v", ty)
```
At that point, we should probably return an error to the client. Without it client potentially will wait for a response that will never come.

jakule commented on lib/web/terminal.go:
```diff
@@ -1068,6 +998,199 @@ func (t *WSStream) writeError(err error) error {
+
+			var envelope Envelope
+			if err := proto.Unmarshal(bytes, &envelope); err != nil {
+				t.log.Warnf("Unable to parse message payload %v", err)
```
The same as above.

jakule commented on lib/web/terminal.go:
```diff
@@ -166,6 +170,11 @@ type TerminalHandlerConfig struct {
+	// if the user is not joining a session.
+	Tracker types.SessionTracker
+	// Clock used for presence checking.
+	Clock clockwork.Clock
```
Does the clock need to be public?

jakule commented on lib/client/client.go:
```diff
@@ -1691,7 +1691,7 @@ func (c *NodeClient) handleGlobalRequests(ctx context.Context, requestCh <-chan
 				}
 
-				c.OnMFA()
+				go c.OnMFA()
```
Can you add a comment on why this call is not async?

rosstimothy committed:
refactor envelope handling

rosstimothy committed:
fix build

rosstimothy committed:
fix: revert test debug timeout

rosstimothy committed:
fix: use local context in tests

rosstimothy committed:
simplify closing streams

rosstimothy committed:
generalize waitForOutput to work with an io.Reader

rosstimothy committed:
fix error handling in stream close

zmb3 commented on lib/client/presence.go:
```diff
@@ -23,26 +23,59 @@ import (
+
+// PresenceOptions allows passing optional overrides
+// to RunPresenceTask. Mainly used by tests.
+type PresenceOptions struct {
```
Maybe unexport this to keep the API surface smaller? (Other packages only use the exported `WithXXX` functions and don't need to know about the internal struct)

rosstimothy committed:
unexport PresenceOptions

rosstimothy committed:
Improve waitForOutput to match against output in successive reads
